# 4.4 Data Wrangling & Subsetting

Content List
*Importing libraries
*Import csv/Dataframes
*Data Wrangling
*4.4 tasks

## Importing Libraries


In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

## Import CSV/ Dataframes


In [2]:
df_prod = pd.read_csv(r'C:\Users\kaydo\OneDrive\Desktop\Python\Data\OG Data\products.csv')

In [3]:
df_ords = pd.read_csv(r'C:\Users\kaydo\OneDrive\Desktop\Python\Data\OG Data\orders.csv')

In [4]:
df_dep = pd.read_csv(r'C:\Users\kaydo\OneDrive\Desktop\Python\Data\OG Data\departments.csv')


In [24]:
#establsih path 
path = r'C:\Users\kaydo\OneDrive\Desktop\Python\Data\OG Data'

## Data Wrangling

In [6]:
#Dropping eval_set column from orders.csv
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [7]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [8]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [9]:
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [10]:
#transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [11]:
df_dep_t = df_dep.T

In [12]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [13]:
new_header = df_dep_t.iloc[0]

In [14]:
df_dep_t_new = df_dep_t[1:]

In [15]:
df_dep_t_new.columns= new_header #set the header row as the df header

In [16]:
data_dict = df_dep_t_new.to_dict('index')

In [17]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [18]:
df_prod['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [19]:
df_prod[df_prod['department_id']==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [20]:
df_snacks= df_prod[df_prod['department_id']==19]

In [21]:
df_snacks_2 = df_prod.loc[df_prod['department_id'] == 19]

In [22]:
df_snacks_3 = df_prod.loc[df_prod['department_id'].isin([19])]

In [25]:
df_ords.to_csv(os.path.join(path,'orders_wrangled.csv'))

## Exercise 4.4

In [26]:
#change a numeric variable to a suitable format
df_ords['user_id']=df_ords['user_id'].astype('str')

In [27]:
df_ords['user_id'].dtype

dtype('O')

In [30]:
df_ords.columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [32]:
#change variable in df_ords w/o overwriting data frame
df_ords.rename(columns={'orders_day_of_week':'Order_day_of_week'})

order_id user_id eval_set  order_number  Order_day_of_week  \
0        2539329       1    prior             1                  2   
1        2398795       1    prior             2                  3   
2         473747       1    prior             3                  3   
3        2254736       1    prior             4                  4   
4         431534       1    prior             5                  4   
...          ...     ...      ...           ...                ...   
3421078  2266710  206209    prior            10                  5   
3421079  1854736  206209    prior            11                  4   
3421080   626363  206209    prior            12                  1   
3421081  2977660  206209    prior            13                  1   
3421082   272231  206209    train            14                  6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [34]:
#find the busiest hour for placing orders
df_ords['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

The busiest hour of the day is 10 am with 288418 orders!


In [35]:
#Determine the meaning behind value 4 in department id column using data dictionary
print(data_dict.get('4'))

{'department': 'produce'}


In [36]:
#Find Breakfast items sales by creating a subset
df_breakfast= df_prod[df_prod['department_id']==14]

In [37]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [38]:
#Find all observations from the entire datafrome that includes alcohol,deli,beverages, meat/seafood

In [39]:
#Alcohol
df_alcohal= df_prod[df_prod['department_id']==5]

In [40]:
df_alcohal

product_id                               product_name  aisle_id  \
51             52                        Mirabelle Brut Rose       134   
118           119                     Chardonnay Paso Robles        62   
149           150                                  Brut Rosé       134   
233           234                          Tennessee Whiskey       124   
248           249             Pinot Grigio, California, 2010        62   
...           ...                                        ...       ...   
49548       49544                    Cabernet Sauvignon Wine        28   
49566       49562              Blanc De Noirs Sparkling Wine       134   
49610       49606                        Organic Natural Red        28   
49665       49661                                      Porto       134   
49688       49684  Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
51                 5    14.4  
118                5     5.5  
149                5    12.9  
233                5     3.1  
248                5     2.7  
...              ...     ...  
49548              5     9.8  
49566              5     5.7  
49610              5     6.8  
49665              5     8.2  
49688              5     5.3  

[1056 rows x 5 columns]

In [41]:
#Deli
df_deli= df_prod[df_prod['department_id']==20]

In [42]:
df_deli

product_id                                   product_name  aisle_id  \
48             49  Vegetarian Grain Meat Sausages Italian - 4 CT        14   
61             62        Premium Deli Oven Roasted Turkey Breast        96   
73             74                Artisan Chick'n & Apple Sausage        14   
84             85                           Soppressata Piccante        96   
108           109                         Grape Leaf Hummus Wrap        13   
...           ...                                            ...       ...   
49558       49554                          Roasted Garlic Hommus        67   
49564       49560    Selects Natural Slow Roasted Chicken Breast        96   
49585       49581                   Pinto Bean and Cheese Pupusa        13   
49609       49605                     Classic Hummus Family Size        67   
49642       49638           Organic Superfirm Vacuum Packed Tofu        14   

       department_id  prices  
48                20    10.1  
61                20    14.6  
73                20     3.0  
84                20     8.9  
108               20     7.1  
...              ...     ...  
49558             20    14.8  
49564             20    14.5  
49585             20    10.5  
49609             20     3.5  
49642             20     9.0  

[1322 rows x 5 columns]

In [43]:
#Beverages
df_beverages= df_prod[df_prod['department_id']==7]

In [44]:
df_beverages

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
19             20  Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
...           ...                                             ...       ...   
49659       49655                                     Apple Cider        98   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
19                 7     6.0  
...              ...     ...  
49659              7    10.7  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  

[4365 rows x 5 columns]

In [45]:
#Meat/seafood
df_meat_seafood= df_prod[df_prod['department_id']==12]

In [46]:
df_meat_seafood

product_id                                       product_name  \
16             17                                  Rendered Duck Fat   
22             23                             Organic Turkey Burgers   
34             35     Italian Herb Porcini Mushrooms Chicken Sausage   
39             40  Beef Hot Links Beef Smoked Sausage With Chile ...   
83             84                                         Lamb Shank   
...           ...                                                ...   
49425       49421                                   Salame, Peppered   
49440       49436                              Imitation Crab Flakes   
49509       49505                                Hot Italian Sausage   
49655       49651                                       Beef Brisket   
49674       49670                                Beef Summer Sausage   

       aisle_id  department_id  prices  
16           35             12    17.1  
22           49             12     8.2  
34          106             12    15.1  
39          106             12    22.5  
83            7             12    24.3  
...         ...            ...     ...  
49425       106             12    12.2  
49440        15             12    23.5  
49509       106             12    15.8  
49655       122             12    20.7  
49674       106             12    19.2  

[907 rows x 5 columns]

In [49]:
#Create a subset containing the required information
df_dinner_party= df_prod.loc[df_prod['department_id'].isin([5,20,7,12])]

In [50]:
df_dinner_party

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

In [51]:
#df_dinner_party has 7650 rows of data

In [59]:
#Extract all information on user 1
df_ords.head()

order_id user_id eval_set  order_number  Order_day_of_week  \
0  2539329       1    prior             1                  2   
1  2398795       1    prior             2                  3   
2   473747       1    prior             3                  3   
3  2254736       1    prior             4                  4   
4   431534       1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [60]:
df_ords_user1 = df_ords[df_ords['user_id']==1]

In [62]:
df_ords_user1
#I am having an issue with this output. Each time I am writing the code it is not bringing the correct data in

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, Order_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [63]:
#describe user behavior
df_ords_user1.describe()

order_number  Order_day_of_week  order_hour_of_day  \
count           0.0                0.0                0.0   
mean            NaN                NaN                NaN   
std             NaN                NaN                NaN   
min             NaN                NaN                NaN   
25%             NaN                NaN                NaN   
50%             NaN                NaN                NaN   
75%             NaN                NaN                NaN   
max             NaN                NaN                NaN   

       days_since_prior_order  
count                     0.0  
mean                      NaN  
std                       NaN  
min                       NaN  
25%                       NaN  
50%                       NaN  
75%                       NaN  
max                       NaN

In [64]:
#establsih path 2
path_2 = r'C:\Users\kaydo\OneDrive\Desktop\Python\Data\Prepared Data'

In [65]:
# Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder
df_ords.to_csv(os.path.join(path,'orders_wrangled.csv'))

In [66]:
#Export df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder 
df_dep_t_new.to_csv(os.path.join(path,'departments_wrangled.csv'))